### PR TITLE
Added `py3dns` recipe

### PR DESCRIPTION
**Pull request by Neizvestnyj**
Same as in the [android](https://github.com/kivy/python-for-android/pull/2590), but I change localhost to **8.8.8.8**

```py
from kivy.app import App
from kivy.uix.button import Button

import DNS
import smtplib


class TestApp(App):
    def build(self):
        return Button(text='Get DNS',
                      on_release=lambda *args: print(self.get_dns_host('gmail.com')),
                      )

    @staticmethod
    def get_dns_host(hostname: str, timeout=5):
        valid_host = None
        try:
            DNS.DiscoverNameServers()
            mx_hosts = DNS.mxlookup(hostname, timeout)
            for mx in mx_hosts:
                smtp = smtplib.SMTP()
                try:
                    smtp.connect(mx[1])
                except smtplib.SMTPConnectError:
                    continue  # try the next MX server in list
                else:
                    valid_host = mx[1]
        except Exception as err:
            print(err)

        return valid_host


TestApp().run()
```

Screenshot 2022-07-09 at 00 07 33

### CODE REVIEW

**Review by misl6 — COMMENTED** [review]
Let's wait for the CI to do its jobs, but overall looks good.
I'm only worried about forcing users to use `8.8.8.8` without warning them.

Did you checked for alternate solutions? (E.g. Like exposing an API via the `ios` module?)

If we decide that using the Google DNS is the only feasible and easy way to do it, how about including ´8.8.4.4´ as a fallback?

**Comment by Neizvestnyj** [comment]
> 



> Let's wait for the CI to do its jobs, but overall looks good. I'm only worried about forcing users to use `8.8.8.8` without warning them.
> 
> Did you checked for alternate solutions? (E.g. Like exposing an API via the `ios` module?)
> 
> If we decide that using the Google DNS is the only feasible and easy way to do it, how about including ´8.8.4.4´ as a fallback?

I dont now how to use ios module for that, but using 8.8.4.4 is a good solution (and easy). For exmaple we try to check 8.8.8.8 using socket, if it does not be used, we switch to 8.8.4.4

**Review by misl6 — CHANGES_REQUESTED** [review]
The CI noticed an issue on PEP8 checks, can you take care of it?

**Comment by Neizvestnyj** [comment]
> The CI noticed an issue on PEP8 checks, can you take care of it?

yes, I make new commit with some patch fixes

**Comment by Neizvestnyj** [comment]
@misl6 How about this check script?

```py
import socket

def check_host(host="8.8.8.8", port=53, timeout=3):
    """
    Host: 8.8.8.8 (google-public-dns-a.google.com)
    OpenPort: 53/tcp
    Service: domain (DNS/TCP)
    """
    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except socket.error as ex:
        print(ex)
        return False

host = "8.8.8.8"
if not check_host(host):
	host = "8.8.4.4"
```

**Comment by misl6** [comment]
> @misl6 How about this check script?
> 
> ```python
> import socket
> 
> def check_host(host="8.8.8.8", port=53, timeout=3):
>     """
>     Host: 8.8.8.8 (google-public-dns-a.google.com)
>     OpenPort: 53/tcp
>     Service: domain (DNS/TCP)
>     """
>     try:
>         socket.setdefaulttimeout(timeout)
>         socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
>         return True
>     except socket.error as ex:
>         return False
> 
> host = "8.8.8.8"
> if not check_host(host):
> 	host = "8.8.4.4"
> ```

Isn't `defaults['server'] = ['8.8.8.8', '8.8.4.4']` instead of `defaults['server'].append('8.8.8.8')` enough?

**Comment by Neizvestnyj** [comment]
> ['8.8.8.8', '8.8.4.4']

Of course, I dont think about it :)

**Comment by Neizvestnyj** [comment]
@misl6 Done

File: kivy_ios/recipes/py3dns/ios.patch [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ from . import Type,Class,Opcode
+ import asyncore
++import os
+ #
+ # This random generator is used for transaction ids and port selection.  This
+ # is important to prevent spurious results from lost packets, and malicious
+ # cache poisoning.  This doesn't matter if you are behind a caching nameserver
+ # or your app is a primary DNS server only. To install your own generator,
+-# replace DNS.Base.random.  SystemRandom uses /dev/urandom or similar source.  
++# replace DNS.Base.random.  SystemRandom uses /dev/urandom or similar source.
```
Comment (misl6) [inline kivy_ios/recipes/py3dns/ios.patch:16]: Why does that line need to be patched?

File: kivy_ios/recipes/py3dns/ios.patch [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
++        return
+ 
+ def ParseResolvConfFromIterable(lines):
+     "parses a resolv.conf formatted stream and sets defaults for name servers"
+@@ -61,7 +66,7 @@ def ParseResolvConfFromIterable(lines):
+         if not line or line[0]==';' or line[0]=='#':
+             continue
+         fields=line.split()
+-        if len(fields) < 2: 
++        if len(fields) < 2:
```
Comment (misl6) [inline kivy_ios/recipes/py3dns/ios.patch:40]: Why does that line need to be patched?

File: kivy_ios/recipes/py3dns/ios.patch [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+             continue
+         if fields[0]=='domain' and len(fields) > 1:
+             defaults['domain']=fields[1]
+@@ -212,7 +217,7 @@ class DnsRequest:
+                 source_port = random.randint(1024,65535)
+                 self.s.bind(('', source_port))
+                 break
+-            except socket.error as msg: 
++            except socket.error as msg:
```
Comment (misl6) [inline kivy_ios/recipes/py3dns/ios.patch:49]: Why does that line need to be patched?

**Review by misl6 — CHANGES_REQUESTED** [review]
Ouch. Looks like the patch contains some extra un-needed lines.
Can you take care of it?

File: kivy_ios/recipes/py3dns/ios.patch (hunk 1 above)
Comment (Neizvestnyj) [reply kivy_ios/recipes/py3dns/ios.patch:16]: I think PyCharm make some background fixed

Comment (misl6) [reply kivy_ios/recipes/py3dns/ios.patch:16]: Ok, consider keeping `patch` files as small as possible, without including any extra un-needed change.
Let me know when the patch is cleaned up.

Comment (Neizvestnyj) [reply kivy_ios/recipes/py3dns/ios.patch]: Done

File: kivy_ios/recipes/py3dns/ios.patch [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ #
+ # This random generator is used for transaction ids and port selection.  This
+ # is important to prevent spurious results from lost packets, and malicious
+@@ -50,8 +51,12 @@ defaults= { 'protocol':'udp', 'port':53, 'opcode':Opcode.QUERY,
+ 
+ def ParseResolvConf(resolv_path="/etc/resolv.conf"):
+     "parses the /etc/resolv.conf file and sets defaults for name servers"
+-    with open(resolv_path, 'r') as stream:
+-        return ParseResolvConfFromIterable(stream)
++    if os.path.existst(resolv_path):
```
Comment (misl6) [inline kivy_ios/recipes/py3dns/ios.patch:19]: Looks you did a typo, as `os.path.existst` is not a valid method

**Review by misl6 — CHANGES_REQUESTED** [review]

File: kivy_ios/recipes/py3dns/ios.patch (hunk 4 above)
Comment (Neizvestnyj) [reply kivy_ios/recipes/py3dns/ios.patch]: After the first commit, I made some changes and for some reason decided to change this line as well :), thank you

**Review by misl6 — CHANGES_REQUESTED** [review]
The automated tests on changed recipes are now failing due to a malformed patch.
```
2022-07-10T08:59:48.6004950Z   STDERR:
2022-07-10T08:59:48.6005440Z /usr/bin/patch: **** malformed patch at line 13: @@ -50,8 +51,12 @@ defaults= { 'protocol':'udp', 'port':53, 'opcode':Opcode.QUERY,
```

Can you test it locally and check the patch?

**Comment by Neizvestnyj** [comment]
> The automated tests on changed recipes are now failing due to a malformed patch.
> 
> ```
> 2022-07-10T08:59:48.6004950Z   STDERR:
> 2022-07-10T08:59:48.6005440Z /usr/bin/patch: **** malformed patch at line 13: @@ -50,8 +51,12 @@ defaults= { 'protocol':'udp', 'port':53, 'opcode':Opcode.QUERY,
> ```
> 
> Can you test it locally and check the patch?

I update path file.
I download this and unpack https://launchpad.net/py3dns/trunk/3.2.1/+download/py3dns-3.2.1.tar.gz
```bash
cd py3dns-3.2.1
git init
git add .
git commit -m "test"
# make changes in file
git diff > ios.patch
```

**Review by misl6 — APPROVED** [review]
Thank you!

**Pull request merged**